### PR TITLE
Refactoring of tools

### PR DESCRIPTION
**Pull request by patriksvensson**
Added new tool class that requires it's settings to have a tool path.
Obsoleted the old tool class so it can be used for a little while longer.

Also started refactoring the tests related to tools to make them easier to understand and write.

### CODE REVIEW

**Comment by gep13** [comment]
Oh wow, you have been busy! :smile_cat: 

When are you planning on pulling this in?  If/when you do, I can update my existing PR's to use it as well.  I can use the changes that you made to the Chocolatey Addin as a guide.

:+1:

**Comment by patriksvensson** [comment]
@gep13 Hopefully this will be ready before tomorrow. Took the opportunity to refactor the tool tests as well to make them easier to understand/write now when they have more in common.

**Comment by gep13** [comment]
@patriksvensson yip, I was reading through the changes in this PR, and I could see the optimizations, great work!

**Comment by patriksvensson** [comment]
Ok, took longer than I thought to do this. Anyway, only three tools left now and this can be merged. Got the hang of it so it shouldn't take too long to do this. Expect this to be finished tomorrow :smile:

**Comment by gep13** [comment]
Quality! :+1: Let me know as/when you get this pulled in, and I will start updating my existing PR's.

File: src/Cake.Common.Tests/Fixtures/Tools/Chocolatey/ChocolateyFixture.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    internal abstract class ChocolateyFixture<TSettings> : ChocolateyFixture<TSettings, ToolFixtureResult>
+        where TSettings : ToolSettings, new()
+    {
+        protected override ToolFixtureResult CreateResult(FilePath toolPath, ProcessSettings process)
+        {
+            return new ToolFixtureResult(toolPath, process);
+        }
+    }
+
+    internal abstract class ChocolateyFixture<TSettings, TFixtureResult> : ToolFixture<TSettings, TFixtureResult>
```
Comment (devlead) [inline src/Cake.Common.Tests/Fixtures/Tools/Chocolatey/ChocolateyFixture.cs]: Trailing space.

File: src/Cake.Common.Tests/Fixtures/Tools/Chocolatey/ChocolateyFixture.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        where TSettings : ToolSettings, new()
+    {
+        protected override ToolFixtureResult CreateResult(FilePath toolPath, ProcessSettings process)
+        {
+            return new ToolFixtureResult(toolPath, process);
+        }
+    }
+
+    internal abstract class ChocolateyFixture<TSettings, TFixtureResult> : ToolFixture<TSettings, TFixtureResult> 
+        where TSettings : ToolSettings, new()
```
Comment (devlead) [inline src/Cake.Common.Tests/Fixtures/Tools/Chocolatey/ChocolateyFixture.cs]: Trailing space.

File: src/Cake.Common.Tests/Fixtures/Tools/GitReleaseManagerFixture.cs [hunk 3]
```diff
@@ -1,68 +1,21 @@
...
 using Cake.Core.Diagnostics;
-using Cake.Core.IO;
-using Cake.Testing;
+using Cake.Core.Tooling;
 using NSubstitute;
 
 namespace Cake.Common.Tests.Fixtures.Tools
 {
-    public abstract class GitReleaseManagerFixture
+    internal abstract class GitReleaseManagerFixture<TSettings> : ToolFixture<TSettings>
```
Comment (devlead) [inline src/Cake.Common.Tests/Fixtures/Tools/GitReleaseManagerFixture.cs]: Trailing white space.

**Comment by gep13** [comment]
@devlead :+1: looks like @patriksvensson isn't using the Visual Studio Extension that I recommended :smile:

**Comment by devlead** [comment]
To large to diff the rest here are some more white space issues :)
https://github.com/patriksvensson/cake/blob/f/tool-refactoring/src/Cake.Common.Tests/Unit/Tools/NUnit/NUnitRunnerTests.cs#L189
https://github.com/patriksvensson/cake/blob/f/tool-refactoring/src/Cake.Common.Tests/Unit/Tools/NUnit/NUnitRunnerTests.cs#L212
https://github.com/patriksvensson/cake/blob/f/tool-refactoring/src/Cake.Core/Tooling/Tool.cs#L71
https://github.com/patriksvensson/cake/blob/f/tool-refactoring/src/Cake.Core/Tooling/Tool.cs#L73

Reviewed tests, looks great, now it's allot identical code that I think we step 2 post this PR can remove even more code.

Looking at Common/Core now.

**Comment by devlead** [comment]
@patriksvensson this looks really solid! :+1:  Really love to see all those toolpaths go away(into a common class :) )
Haven't found any major issues, except cosmetics mentioned earlier. Would like to get this merged and up to MyGet so I can test it in action on a projects to make sure no regressions.

Epic work :+1:

**Comment by jrnail23** [comment]
@patriksvensson, you're targeting this for Cake v0.5.6, correct?  When do you anticipate releasing that version?

**Comment by gep13** [comment]
@patriksvensson when are you planning on pulling this in?  Looks like I might have some time on Friday night, so would look to start refactoring my open PR's to use this new functionality if it is pulled in before then :smile:

**Comment by patriksvensson** [comment]
@gep13 We try to not merge our own PRs so it's up to @devlead, but soon I hope :)

**Comment by gep13** [comment]
@patriksvensson gotcha! :+1:

**Comment by patriksvensson** [comment]
@jrnail23 Not sure. Still some stuff in the current milestone to complete.

**Comment by gep13** [comment]
@devlead WHY HAVEN'T YOU MERGED THIS YET! :stuck_out_tongue:

**Comment by patriksvensson** [comment]
@gep13 @devlead lol :)

**Comment by devlead** [comment]
Rerunning Travis builds and local build now ;)

**Comment by devlead** [comment]
@gep13 IT'S ALIVE!

**Comment by gep13** [comment]
Yip, yip, yip.

I am due to go for a 12km run at lunch time today, and heading out to the cinema tonight, but I plan to start rebasing my PR's tomorrow, and refactoring their implementation based on the work that @patriksvensson has done.

**Comment by devlead** [comment]
@gep13 also any tool based add-in should be able to start testing using the pre-release MyGet package.

**Comment by gep13** [comment]
@devlead yip, Cake.Gitter shouldn't be affected, but I will have to update Cake.ReSharperReports to take advantage of this as well.

**Pull request merged**